crates/accelerate/src/linear_matrix.rs

```diff
@@ -0,0 +1,104 @@
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ndarray::{s, ArrayViewMut2, Axis};
+use numpy::PyReadwriteArray2;
+use pyo3::prelude::*;
+use rayon::prelude::*;
+
+// Perform ROW operation on a matrix mat
+fn _row_op(mat: &mut ArrayViewMut2<bool>, ctrl: usize, trgt: usize) {
+    let row0 = mat.row(ctrl).to_owned();
+    let mut row1 = mat.row_mut(trgt);
+    row1.zip_mut_with(&row0, |x, &y| *x ^= y);
+}
+
+// Perform COL operation on a matrix mat
+fn _col_op(mat: &mut ArrayViewMut2<bool>, ctrl: usize, trgt: usize) {
+    let col0 = mat.column(ctrl).to_owned();
+    let mut col1 = mat.column_mut(trgt);
+    col1.zip_mut_with(&col0, |x, &y| *x ^= y);
+}
+
+// Gauss elimination of a matrix mat with m rows and n columns.
+// If full_elim = True, it allows full elimination of mat[:, 0 : ncols]
+// Returns the matrix mat.
+fn gauss_elimination(mut mat: ArrayViewMut2<bool>, ncols: Option<usize>, full_elim: Option<bool>) {
+    let (m, mut n) = (mat.nrows(), mat.ncols()); // no. of rows and columns
+    if let Some(ncols_val) = ncols {
+        n = usize::min(n, ncols_val); // no. of active columns
+    }
+
+    let mut r = 0; // current rank
+    let k = 0; // current pivot column
+    let mut new_k = 0;
+    while (r < m) && (k < n) {
+        let mut is_non_zero = false;
+        let mut new_r = r;
+        for j in k..n {
+            new_k = k;
+            for i in r..m {
+                if mat[(i, j)] {
+                    is_non_zero = true;
+                    new_k = j;
+                    new_r = i;
+                    break;
+                }
+            }
+            if is_non_zero {
+                break;
+            }
+        }
+        if !is_non_zero {
+            return; // A is in the canonical form
+        }
+
+        if new_r != r {
+            let temp_r = mat.slice_mut(s![r, ..]).to_owned();
+            let temp_new_r = mat.slice_mut(s![new_r, ..]).to_owned();
+            mat.slice_mut(s![r, ..]).assign(&temp_new_r);
+            mat.slice_mut(s![new_r, ..]).assign(&temp_r);
+        }
+
+        // Copy source row to avoid trying multiple borrows at once
+        let row0 = mat.row(r).to_owned();
+        mat.axis_iter_mut(Axis(0))
+            .into_par_iter()
+            .enumerate()
+            .filter(|(i, row)| {
+                (full_elim == Some(true) && (*i < r) && row[new_k])
+                    || (*i > r && *i < m && row[new_k])
+            })
+            .for_each(|(_i, mut row)| {
+                row.zip_mut_with(&row0, |x, &y| *x ^= y);
+            });
+
+        r += 1;
```

Comment:
Besides the open question around whether parallelism actually improves performance here and also the open question about the an iterator vs a for loop too. If we do end up using parallelism we should add a guard here to prevent running in multiple threads at the same time qiskit is using multiple processes. We don't want to spawn a threadpool if Qiskit is running in multiple processes.

You can see the pattern we use for this in other places: https://github.com/Qiskit/qiskit/blob/main/crates/accelerate/src/pauli_exp_val.rs#L186-L192 

The key is to check `getenv_use_multiple_threads()` to get a boolean value based on Qiskit's environment variables whether we're in a multiprocessing context currently or not. If we are then we should not include the `into_par_iter()` otherwise it's safe to do.

Reply:
Indeed, after benchmarking the code on a linux VM with 8 cores, it seems that doing multiple threading only makes it worse, so I removed it for now.

Original python code:
```
num_qubits = 50, time = 0.02
num_qubits = 100, time = 0.06
num_qubits = 250, time = 0.38
num_qubits = 500, time = 1.68
num_qubits = 1000, time = 8.08
num_qubits = 2500, time = 83.79
```
rust code (single thread):
```
num_qubits = 50, time = 0.00
num_qubits = 100, time = 0.00
num_qubits = 250, time = 0.03
num_qubits = 500, time = 0.20
num_qubits = 1000, time = 2.18
num_qubits = 2500, time = 29.95
```
rust code (multiple threads):
```
num_qubits = 50, time = 0.30
num_qubits = 100, time = 0.53
num_qubits = 250, time = 2.75
num_qubits = 500, time = 8.72
num_qubits = 1000, time = 22.15
num_qubits = 2500, time = 140.68 
```